### PR TITLE
use select() for main loop

### PR DESCRIPTION
**Pull request by mb-gh**
This is the general idea...
May need some refinement.

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -185,12 +253,51 @@ int main (int argc, char** argv)
...
+					it != pc_loop_fd_list.end(); ++it)
+				{
+					int const fd = *it;
+					nfds = std::max(nfds, *it + 1);
+					FD_SET(fd, &pc_loop_fd_set);
+				}
+				timeval timeout = {refreshdelay, 0};
+				if( select(nfds, &pc_loop_fd_set, 0, 0, &timeout) == -1 )
+				{
+					perror("error in select()\n");
```
Comment (raboof) [inline main.cpp]: the `perror` error message doesn't need to be ended with a `\n`

File: main.cpp [hunk 2]
```diff
@@ -98,13 +131,29 @@ int main (int argc, char** argv)
...
 	if (NEEDROOT && (geteuid() != 0))
 		forceExit(false, "You need to be root to run NetHogs!");
+	
+	fd_set pc_loop_fd_set;
+	std::vector<int> pc_loop_fd_list;
+	bool pc_loop_use_select = true;	
+	
+	if( pc_loop_use_select )
+	{
+		pc_loop_fd_list.push_back(self_pipe.first);
```
Comment (raboof) [inline main.cpp]: Is this supposed to be here? It's also present further down, where it makes more sense :)

File: main.cpp [hunk 3]
```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
...
 	}
+	
+	//close file descriptors
+	for(std::vector<int>::const_iterator it=pc_loop_fd_list.begin();
+		it != pc_loop_fd_list.end(); ++it)
+	{
+		close(*it);
+	}
+	
+	procclean();
```
Comment (raboof) [inline main.cpp]: Taking this out of the callback and moving it to right after the main loop is much nicer :+1:

File: main.cpp [hunk 4]
```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
...
+			else
+			{
+				// If select() not possible, pause to prevent 100%
+				// Pause 10 milliseconds
+				usleep(1000);
+			}
 		}
 	}
+	
+	//close file descriptors
```
Comment (raboof) [inline main.cpp]: Perhaps we should do all the cleanups in a separate helper function?

File: main.cpp [hunk 5]
```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
...
+				{
+					std::cout << "exited by select\n";
+					//exit the loop
+					break;
+				}
+			}
+			else
+			{
+				// If select() not possible, pause to prevent 100%
+				// Pause 10 milliseconds
```
Comment (raboof) [inline main.cpp]: This is actually 1 millisecond, right? I think we can remove this comment.

File: main.cpp [hunk 6]
```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
...
 				ui_tick();
 			}
 			do_refresh();
-			needrefresh = false;
 		}
-
-		// If no packets were read at all this iteration, pause to prevent 100%
-		// CPU utilisation;
+ 
+		//if not packets, do a select() until next packet
```
Comment (raboof) [inline main.cpp]: Hmm, don't we want to do a select() regardless of whether we read packets? Probably doesn't matter much in practice though.

File: main.cpp [hunk 7]
```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
...
 		if (!packets_read)
 		{
-			usleep(100);
+			if( pc_loop_use_select )
```
Comment (raboof) [inline main.cpp]: perhaps move this to a 'wait for next trigger' helper function?

File: main.cpp [hunk 8]
```diff
@@ -1,4 +1,9 @@
 #include "nethogs.cpp"
+#include <fcntl.h>
+#include <vector>
+
+std::pair<int,int> self_pipe = std::make_pair(-1, -1);
```
Comment (raboof) [inline main.cpp]: The reason for this pipe is to be able to trigger the select loop from the interrupt handler, right? Might be worth adding a comment about that.

Comment (mb-gh) [inline main.cpp]: OK

File: main.cpp (hunk 2 above)
Comment (mb-gh) [inline main.cpp]: Yes indeed. once is enough

File: main.cpp (hunk 7 above)
Comment (mb-gh) [inline main.cpp]: OK

File: main.cpp (hunk 6 above)
Comment (mb-gh) [inline main.cpp]: not sure, if it may block waiting for new packets ignoring the ones already waiting

**Comment by mb-gh** [comment]
It should be better now.
I am calling it a day today.
If you find other issues, Tomorrow I will have some time.

**Comment by raboof** [comment]
Lovely! Some stylistic inconsistencies but we'll fix those with #44 later :)

**Pull request merged**